File: src/dailyprogrammer/easy/_372/PerfectlyBalanced.java
package dailyprogrammer.easy._372;

import java.util.HashMap;

public class PerfectlyBalanced {
    static final String[] INPUT = {
        "xxxyyyzzz", "abccbaabccba", "xxxyyyzzzz", 
        "abcdefghijklmnopqrstuvwxyz", "pqq", "fdedfdeffeddefeeeefddf",
        "www", "x", ""
    };
    public static void main(String[] args) {
        for (int i = 0; i < INPUT.length; i++) {
            System.out.printf("balanced_bonus(\"%s\") => %s\n", INPUT[i], 
                balanced_bonus(INPUT[i]));
        }
    }

    public static HashMap<String, Integer> count_letters(String s) {
        /*
         * destructive to String s, cuts it down
         */
        HashMap<String, Integer> hm = new HashMap<String, Integer>();
        while (s.length() > 0) {
            // cut off front
            String ch = s.substring(0, 1);
            s = s.substring(1);
            if (!hm.containsKey(ch))
                hm.put(ch, 1);
            else
                hm.put(ch, hm.get(ch) + 1);
        }
        return hm;
    }

    public static boolean balanced_bonus(String s) {
        HashMap<String, Integer> hm = count_letters(s);
        // handle 0-1 letter cases
        if (hm.size() <= 1)
            return true;
        // handle 2+ letter cases
        int i = 0;
        Integer first_count = 0;
        for (Integer count : hm.values()) {
            if (i == 0)  // annoying to get first hm value out of hm
                first_count = count;
            if (first_count != count)
                return false;
            i++;
        }
        return true;
    }
}

File: src/dailyprogrammer/easy/_372/PerfectlyBalanced.py
INPUT = ['xxxyyyzzz', 'abccbaabccba', 'xxxyyyzzzz', 
        'abcdefghijklmnopqrstuvwxyz', 'pqq', 'fdedfdeffeddefeeeefddf',
        'www', 'x', '']

def count_letters() -> dict:
    letter_dict = dict()
    for ch in s:
        if ch not in letter_dict:
            letter_dict[ch] = 1
        else:
            letter_dict[ch] += 1
    return letter_dict

def balanced_bonus(s: str) -> bool:
    """steps
    1) count_letter() : count letters and store in dict
    2) handle 0 or 1 letter
    3) handle 2 letter and above
    """
    letter_dict = count_letters()
    if len(letter_dict) <= 1:  # 0 or 1 letter
        return True
    last_c = letter_dict.popitem()[1]
    for count in letter_dict.values():
        if last_c != count:
            return False
    return True

def main():
    for i in range(0, len(INPUT), +1):
        print('balanced_bonus("%s") => %s' % \
            (INPUT[i], balanced_bonus(INPUT[i])))

if __name__ == '__main__':
    main()